File: src/main/org/nlogo/awt/DarkenImageFilter.java
package org.nlogo.awt;

public strictfp class DarkenImageFilter
    extends java.awt.image.RGBImageFilter {

  private static final int OPAQUE_ALPHA = 0xff000000;

  private final double level;

  public DarkenImageFilter(double level) {
    this.level = level;
    canFilterIndexColorModel = true;
  }

  @Override
  public int filterRGB(int x, int y, int rgb) {
    int red = (rgb >> 16) & 0xff;
    int green = (rgb >> 8) & 0xff;
    int blue = (rgb) & 0xff;
    red = (int) ((1.0 - level) * red);
    green = (int) ((1.0 - level) * green);
    blue = (int) ((1.0 - level) * blue);
    int newRGB = rgb & OPAQUE_ALPHA;
    newRGB = newRGB | (red << 16);
    newRGB = newRGB | (green << 8);
    return newRGB | (blue);
  }

}

File: src/main/org/nlogo/awt/DarkenImageFilter.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.awt

class DarkenImageFilter(level: Double)
extends java.awt.image.RGBImageFilter {
  private final val OpaqueAlpha = 0xff000000
  canFilterIndexColorModel = true
  override def filterRGB(x: Int, y: Int, rgb: Int): Int = {
    var red = (rgb >> 16) & 0xff
    var green = (rgb >> 8) & 0xff
    var blue = (rgb) & 0xff
    red = ((1.0 - level) * red).toInt
    green = ((1.0 - level) * green).toInt
    blue = ((1.0 - level) * blue).toInt
    (rgb & OpaqueAlpha) | (red << 16) | (green << 8) | blue
  }
}
